src/Raven.Server/Documents/TcpHandlers/SubscriptionConnectionForShard.cs

```diff
@@ -36,11 +48,67 @@ protected override DynamicJsonValue AcceptMessage()
 
     protected override RawDatabaseRecord GetRecord(TransactionOperationContext context) => _serverStore.Cluster.ReadRawDatabaseRecord(context, ShardName);
 
+    protected override string SetLastChangeVectorInThisBatch(IChangeVectorOperationContext context, string currentLast, Document sentDocument)
+    {
+        if (sentDocument.Etag == 0) // got this document from resend
+            return currentLast;
+
+        var vector = context.GetChangeVector(sentDocument.ChangeVector);
+
+        var result = ChangeVectorUtils.MergeVectors(
+            currentLast,
+            ChangeVectorUtils.NewChangeVector(_serverStore.NodeTag, sentDocument.Etag, _shardedDatabase.DbBase64Id),
+            vector.Order);
+
+        return result;
+    }
+
+    protected override async Task UpdateStateAfterBatchSentAsync(IChangeVectorOperationContext context, string lastChangeVectorSentInThisBatch)
+    {
+        var vector = context.GetChangeVector(lastChangeVectorSentInThisBatch);
+        vector.TryRemoveIds(_dbIdToRemove, context, out vector);
+
+        await base.UpdateStateAfterBatchSentAsync(context, vector.Order);
+
+        var p = Processor as ShardedDocumentsDatabaseSubscriptionProcessor;
+        if (p.Skipped != null)
+        {
+            for (int i = CurrentBatch.Count - 1; i >= 0; i--)
+            {
+                var item = CurrentBatch[i];
+                if (p.Skipped.Contains(item.Document.Id))
+                {
+                    CurrentBatch.RemoveAt(i);
+                    item.Document.Dispose();
+                }
+            }
+        }
+    }
+
+    protected override bool FoundAboutMoreDocs()
+    {
+        if (base.FoundAboutMoreDocs())
+            return true;
+
+        if (_state.HasDocumentFormResendForShard())
+            return true;
+
+        return false;
+    }
+
     public SubscriptionConnectionsState GetSubscriptionConnectionStateForShard()
     {
         var subscriptions = TcpConnection.DocumentDatabase.SubscriptionStorage.Subscriptions;
         var state = subscriptions.GetOrAdd(SubscriptionId, subId => new SubscriptionConnectionsStateForShard(DatabaseName, subId, TcpConnection.DocumentDatabase.SubscriptionStorage));
         _subscriptionConnectionsState = state;
+        _state = state as SubscriptionConnectionsStateForShard;
```

Comment:
are we expecting here null? maybe avoid safe cast or add a Debug.Assert at least?

Reply:
done